File: app/src/main/java/com/czbix/v2ex/model/Avatar.java
package com.czbix.v2ex.model;

import android.os.Parcel;
import android.os.Parcelable;
import androidx.annotation.NonNull;

import com.czbix.v2ex.AppCtx;
import com.czbix.v2ex.common.exception.FatalException;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;

import java.util.concurrent.ExecutionException;
import java.util.regex.Pattern;

public class Avatar implements Parcelable {
    public static final float DENSITY;
    public static final int SIZE_LARGE = 73;
    public static final int SIZE_NORMAL = 48;
    public static final int SIZE_MINI = 24;

    private static final Pattern PATTERN = Pattern.compile("(mini|normal|large)");
    private static final String LARGE = "large";
    private static final String NORMAL = "normal";
    private static final String MINI = "mini";

    private final String mBaseUrl;

    static {
        DENSITY = AppCtx.getInstance().getResources().getDisplayMetrics().density;
    }

    Avatar(String baseUrl) {
        mBaseUrl = baseUrl;
    }

    public String getLarge() {
        return getUrl(mBaseUrl, LARGE);
    }

    public String getNormal() {
        return getUrl(mBaseUrl, NORMAL);
    }

    public String getMini() {
        return getUrl(mBaseUrl, MINI);
    }

    public String getBaseUrl() {
        return mBaseUrl;
    }

    public String getUrlByDp(float dp) {
        final float size = dp * DENSITY;
        return getUrlByPx(size);
    }

    public String getUrlByPx(float size) {
        if (size <= SIZE_MINI) {
            return getMini();
        } else if (size <= SIZE_NORMAL) {
            return getNormal();
        } else {
            return getLarge();
        }
    }

    public static String getUrl(String baseUrl, String size) {
        return "https:" + String.format(baseUrl, size);
    }

    private static String getBaseUrl(String url) {
        return PATTERN.matcher(url).replaceFirst("%s");
    }

    public static class Builder {
        private static final LoadingCache<String, Avatar> CACHE;
        private String mBaseUrl;

        static {
            CACHE = CacheBuilder.newBuilder()
                    .softValues()
                    .initialCapacity(32)
                    .maximumSize(128)
                    .build(new CacheLoader<String, Avatar>() {
                        @Override
                        public Avatar load(@NonNull String key) throws Exception {
                            return new Avatar(key);
                        }
                    });
        }

        public Builder setUrl(String url) {
            mBaseUrl = getBaseUrl(url);
            return this;
        }

        public Builder setBaseUrl(String url) {
            mBaseUrl = url;
            return this;
        }

        public Avatar createAvatar() {
            try {
                return CACHE.get(mBaseUrl);
            } catch (ExecutionException e) {
                throw new FatalException(e);
            }
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.mBaseUrl);
    }

    protected Avatar(Parcel in) {
        this.mBaseUrl = in.readString();
    }

    public static final Creator<Avatar> CREATOR = new Creator<Avatar>() {
        public Avatar createFromParcel(Parcel source) {
            return new Avatar(source);
        }

        public Avatar[] newArray(int size) {
            return new Avatar[size];
        }
    };
}

File: app/src/main/kotlin/com/czbix/v2ex/model/Avatar.kt
package com.czbix.v2ex.model

import android.os.Parcelable
import androidx.room.Ignore
import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import kotlinx.android.parcel.Parcelize
import java.util.regex.Pattern

@Parcelize
data class Avatar
@JvmOverloads
constructor(
        val baseUrl: String,
        @Ignore
        val isGravatar: Boolean = baseUrl.contains("gravatar")
) : Parcelable {
    fun getUrlByPx(size: Int): String {
        if (isGravatar) {
            return getUrl(baseUrl, size)
        }

        val sizeName = when {
            size <= SIZE_MINI -> MINI
            size <= SIZE_NORMAL -> NORMAL
            else -> LARGE
        }

        return getUrl(baseUrl, sizeName)
    }

    class Builder {
        private lateinit var mBaseUrl: String
        private var isGravatar: Boolean? = null

        fun setUrl(url: String): Builder {
            val pair = parseBaseUrl(url)
            isGravatar = pair.first
            mBaseUrl = pair.second
            return this
        }

        fun setBaseUrl(url: String): Builder {
            mBaseUrl = url
            isGravatar = null
            return this
        }

        fun build(): Avatar {
            return CACHE.get(mBaseUrl) {
                if (isGravatar != null && isGravatar!!) {
                    Avatar(mBaseUrl, true)
                } else {
                    Avatar(mBaseUrl)
                }
            }
        }

        companion object {
            private val CACHE: Cache<String, Avatar>

            init {
                CACHE = CacheBuilder.newBuilder()
                        .softValues()
                        .initialCapacity(32)
                        .maximumSize(128)
                        .build()
            }
        }
    }

    companion object {
        const val SIZE_LARGE = 73
        const val SIZE_NORMAL = 48
        const val SIZE_MINI = 24

        private val OLD_PATTERN = Pattern.compile("(mini|normal|large)")
        private val PATTERN = Pattern.compile("""(?<=s=)\d+""")

        private const val LARGE = "large"
        private const val NORMAL = "normal"
        private const val MINI = "mini"

        fun getUrl(baseUrl: String, size: String): String {
            return "https:" + String.format(baseUrl, size)
        }

        fun getUrl(baseUrl: String, size: Int): String {
            return "https:" + String.format(baseUrl, size)
        }

        private fun parseBaseUrl(url: String): Pair<Boolean, String> {
            val trimmedUrl = url.substringAfter(':')
            val matcher = PATTERN.matcher(trimmedUrl)
            if (matcher.find()) {
                return true to matcher.replaceFirst("%d")
            }

            return false to OLD_PATTERN.matcher(trimmedUrl).replaceFirst("%s")
        }
    }
}
